flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/BatchCommitterHandler.java

```diff
@@ -33,44 +30,52 @@
  * Runtime {@link org.apache.flink.streaming.api.operators.StreamOperator} for executing {@link
  * Committer} in the batch execution mode.
  *
- * @param <CommT> The committable type of the {@link Committer}.
+ * @param <InputT> The committable type of the {@link Committer}.
  */
-final class BatchCommitterOperator<CommT> extends AbstractStreamOperator<CommT>
-        implements OneInputStreamOperator<CommT, CommT>, BoundedOneInput {
+final class BatchCommitterHandler<InputT, OutputT>
+        extends AbstractCommitterHandler<InputT, OutputT> {
 
     /** Responsible for committing the committable to the external system. */
-    private final Committer<CommT> committer;
+    private final Committer<InputT> committer;
 
-    /** Record all the committables until the end of the input. */
-    private final List<CommT> allCommittables;
+    /**
+     * The committer that is chained to this committer. It's either {@link
+     * GlobalBatchCommitterHandler} or {@link NoopCommitterHandler}.
```

Comment:
Wouldn't is be easier to just not instantiate a `GlobalBatchCommitter` if not necessary and not introduce the `NoopCommitterHandler`.

Reply:
The `NoopCommitterHandler` avoids null checks inside this class. It's a matter of flavor but in general a `NoopCommitterHandler` is the proper OOP approach (also dubbed case-less programming because of VMT).

Reply:
I do not necessarily see the relation to OOP here ;) but isn't it also some kind of optimization to not instantiate this Noop object?

Reply:
Here is some reference https://en.wikipedia.org/wiki/Null_object_pattern TL;DR `null` is an external construct in OOP.

Also don't worry about performance here: First, we have a singleton pattern as an enum that's very efficient. Second, we are not on the hot path.